ambry-store/src/main/java/com.github.ambry.store/HardDeleter.java

```diff
@@ -352,6 +359,24 @@ void preLogFlush() {
   void postLogFlush() {
     /* start token saved before the flush is now safe to be persisted */
     startTokenSafeToPersist = startTokenBeforeLogFlush;
+
+    hardDeleteLock.lock();
+    try {
+      // PersistCleanupToken because startTokenSafeToPersist changed.
+      pruneHardDeleteRecoveryRange();
+      persistCleanupToken();
```

Comment:
What if recovery range has been pruned but failed to persist cleanup token, will it be a problem?

Reply:
It's fine if it's fails to persist, then we will wait for the next one. 
The assumption is that the file on the disk contains a valid range and all the records between this range. And every time we change the file (prune and then hard delete), we replace the old range with a new one. If we fail to persist it, then next time when we restart, we are going to recover from a older range, which is totally fine, since the hard delete operation is idempotent.